src/adr/0034-distributed-listeners-configuration.md

```diff
@@ -0,0 +1,106 @@
+# 34. Distributed Mailbox Listeners Configuration
+
+Date: 2020-04-23
+
+## Status
+
+Proposed
+
+Supercedes [26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md)
+
+## Context
+
+James enables a user to register additional mailbox listeners.
+
+The distributed James server is handling mailbox event processing (mailboxListener execution) using a RabbitMQ work-queue
+per listener.
+
+Currently, mailbox listeners are determined by the guice bindings of the server and additional mailbox listeners defined
+via configuration files.
+
+While the configuration might be specific for each James server, what actually is defined in RabbitMQ is common. 
+Heterogeneous configuration might then result in unpredictable RabbitMQ resource status. This was left as a limitation
+of [26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md).
+
+## Decision
+
+We need to centralize the definition of mailbox listeners.
+
+An event sourcing system will track the configured mailbox listeners.
+
+It will have the following commands:
+
+ - **AddListener**: Add a given listener. This should be rejected if the group is already used.
+ - **RemoveListener**: Remove a given listener.
+
+It will have the following events:
+
+ - **ListenerAdded**: A mailbox listener is added
+ - **ListenerRemoved**: A mailbox listener is removed
+
+A subscriber will react to these events to modify the RabbitMQ resource accordingly by adding queues, adding or removing
+bindings.
+
+This event sourcing system differs from the one defined in
+[26. Removing a configured additional MailboxListener](0026-removing-configured-additional-mailboxListeners.md) by the
+fact that we should also keep track of listener configuration.
+
+Upon start, James will ensure the **configured mailbox listener event sourcing system** contains the guice injected 
+listeners, and add them if missing (handling the RabbitMQ bindings by this mean), then starts the eventBus which will
+consume the given queues.
+
+If a listener is configured with a class unknown to James, the start-up fails. This can happen if a custom jar, containing
```

Comment:
how do you fix that if the server doesn't start anymore and the expected listener list requires a REST call?

Reply:
1/ By not failing and accepting to start without a listener

2/ By having, in the face of such failure a downgraded mode exposing only the webadmin API to manage groups.

3/ By having a CLI tool for managing groups.

I propose adopting 1/ (because it is simple)  but documenting alternative 2/ and 3/ (that could be seen as enhencements requiring more work)

Reply:
With 1/ you will have servers running while not behaving as configured. The administrator would have to read the logs (or health-check) to figure it out.
Maybe health-check is ok

Reply:
> Maybe health-check is ok

:+1: